src/focus-visible.js

```diff
@@ -230,45 +234,58 @@ function init() {
   }
 
+  // For some kinds of state, we are interested in changes at the global scope
+  // only. For example, global pointer input, global key presses and global
+  // visibility change should affect the state at every scope:
   document.addEventListener('keydown', onKeyDown, true);
   document.addEventListener('mousedown', onPointerDown, true);
   document.addEventListener('pointerdown', onPointerDown, true);
   document.addEventListener('touchstart', onPointerDown, true);
-  document.addEventListener('focus', onFocus, true);
-  document.addEventListener('blur', onBlur, true);
   document.addEventListener('visibilitychange', onVisibilityChange, true);
+
   addInitialPointerMoveListeners();
 
-  document.body.classList.add('js-focus-visible');
+  // For focus and blur, we specifically care about state changes in the local
+  // scope. This is because focus / blur events that originate from within a
+  // shadow root are not re-dispatched from the host element if it was already
+  // the active element in its own scope:
+  scope.addEventListener('focus', onFocus, true);
+  scope.addEventListener('blur', onBlur, true);
+
+  // We detect that a node is a ShadowRoot by ensuring that it is a
+  // DocumentFragment and also has a host property. This check covers native
+  // implementation and polyfill implementation transparently. If we only cared
+  // about the native implementation, we could just check if the scope was
+  // an instance of a ShadowRoot.
+  if (scope.nodeType === Node.DOCUMENT_FRAGMENT_NODE && scope.host) {
+    // Since a ShadowRoot is a special kind of DocumentFragment, it does not
+    // have a root element to add a class to. So, we add this attribute to the
+    // host element instead:
+    scope.host.setAttribute('data-js-focus-visible', '');
+  } else if (scope.nodeType === Node.DOCUMENT_NODE) {
+    document.documentElement.classList.add('js-focus-visible');
+  }
 }
 
-/**
- * Subscription when the DOM is ready
- * @param {Function} callback
- */
-function onDOMReady(callback) {
-  var loaded;
+// Make the polyfill helper globally available. This can be used as a signal to
+// interested libraries that wish to coordinate with the polyfill for e.g.,
+// applying the polyfill to a shadow root:
+window.applyFocusVisiblePolyfill = applyFocusVisiblePolyfill;
 
-  /**
-   * Callback wrapper for check loaded state
-   */
-  function load() {
-    if (!loaded) {
-      loaded = true;
+// Notify interested libraries of the polyfill's presence, in case the polyfill
+// was loaded lazily:
+var event;
 
-      callback();
-    }
-  }
-
-  if (['interactive', 'complete'].indexOf(document.readyState) >= 0) {
-    callback();
-  } else {
-    loaded = false;
-    document.addEventListener('DOMContentLoaded', load, false);
-    window.addEventListener('load', load, false);
-  }
+try {
+  event = new CustomEvent('focus-visible-polyfill-ready');
+} catch (error) {
+  // IE11 does not support using CustomEvent as a constructor directly:
```

Comment:
Does it support `new Event()`?

Reply:
Some very light Googling yields this [Stack Overflow](https://stackoverflow.com/questions/26596123/internet-explorer-9-10-11-event-constructor-doesnt-work) that suggests "no" and "use CustomEvent instead"  😄